### PR TITLE
Pwbasis refactor

### PR DESCRIPTION
**Pull request by antoine-levitt**
This makes more sense and is non-breaking. Based on that we can try to cut down the API further and remove a bunch of constructors...

### CODE REVIEW

**Comment by antoine-levitt** [comment]
(it is technically breaking in the sense that it will not change fft_size for buggy generic FFTs behind your back and will just error, but I think that's fair)

File: src/PlaneWaveBasis.jl [hunk 1]
```diff
@@ -140,24 +150,17 @@ end
...
-    # Validate kpoints and symmetries
-    @assert length(kcoords) == length(ksymops)
-    if symmetries === nothing
-        # TODO instead compute the group generated by with ksymops, or
-        # just retire this constructor. Not critical because this
-        # should not be used in this context anyway...
-        symmetries = vcat(ksymops...)
+                                kcoords::AbstractVector, ksymops,
+                                fft_size, variational, symmetries,
+                                kgrid, kshift, comm_kpts) where {T <: Real}
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:158]: I wonder if we should keep the argument order. Perhaps it's better to first do the generic stuff (`model`, `Ecut`, `fft_size`, `variational`, `symmetries`) then the k-Point stuff (`kcoords`, `ksymops`, `kgrid`, `kshift`, `comm_kpts`).

File: src/PlaneWaveBasis.jl [hunk 2]
```diff
@@ -140,24 +150,17 @@ end
...
+    if !(all(fft_size .== next_working_fft_size.(T, fft_size)))
+        error("Selected fft_size will not work for the buggy generic FFT routines")
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:160]: This error message does not help a user that has never encountered this error and does not know about the internals.

File: src/PlaneWaveBasis.jl [hunk 3]
```diff
@@ -140,24 +150,17 @@ end
...
     end
+    fft_size = Tuple{Int, Int, Int}(fft_size) # need explicit convert from
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:162]: two spaces before `#` (is it that hard to write an emacs macro that fixes that?)

File: src/PlaneWaveBasis.jl [hunk 4]
```diff
@@ -210,16 +210,17 @@ build_kpoints(basis::PlaneWaveBasis, kcoords) =
         "Non-variational calculations are experimental. " *
         "Not all features of DFTK may be supported or work as intended."
     )
-    kpoints = build_kpoints(model, fft_size, kcoords, Ecut; variational=variational)
+    kpoints = build_kpoints(model, fft_size, kcoords_thisproc, Ecut;
+                            variational)
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:214]: here i would skip the linebreak.

File: src/PlaneWaveBasis.jl [hunk 5]
```diff
@@ -231,7 +232,7 @@ build_kpoints(basis::PlaneWaveBasis, kcoords) =
 
     basis = PlaneWaveBasis{T}(
         model, Ecut, variational, kpoints,
-        kweights, ksymops, kgrid, kshift, comm_kpts, krange_thisproc, krange_allprocs,
+        kweights, ksymops_thisproc, kgrid, kshift, kcoords, ksymops, comm_kpts, krange_thisproc, krange_allprocs,
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:235]: but this is definitely too long

File: src/PlaneWaveBasis.jl [hunk 6]
```diff
@@ -245,13 +246,58 @@ build_kpoints(basis::PlaneWaveBasis, kcoords) =
...
 
+# This is the "internal" constructor; the higher-level one below should be preferred
+@timing function PlaneWaveBasis(model::Model{T}, Ecut::Number,
+                                kcoords::AbstractVector, ksymops, symmetries=nothing;
+                                fft_size=nothing, variational=true,
+                                optimize_fft_size=false, supersampling=2,
+                                kgrid=nothing, kshift=nothing,
+                                comm_kpts=MPI.COMM_WORLD,
+                               ) where {T <: Real}
+    mpi_ensure_initialized()
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:257]: Why is this needed?

File: src/PlaneWaveBasis.jl [hunk 7]
```diff
@@ -245,13 +246,58 @@ build_kpoints(basis::PlaneWaveBasis, kcoords) =
...
+Creates a new basis identical to `basis`, but with a custom set of kpoints
 """
 function PlaneWaveBasis(basis::PlaneWaveBasis, kcoords::AbstractVector,
-                        ksymops::AbstractVector, symmetries=nothing)
-    PlaneWaveBasis(basis.model, basis.Ecut, kcoords, ksymops, symmetries;
-                   fft_size=basis.fft_size, variational=basis.variational)
+                        ksymops::AbstractVector, symmetries=vcat(ksymops...))
+    PlaneWaveBasis(basis.model, basis.Ecut, kcoords, ksymops,
+                   basis.fft_size, basis.variational, basis.symmetries,
+                   nothing, nothing, # explicitly given kpoints
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:299]: I think for clarity I would assign these to variables `kgrid` and `kshift` and pass them. That's one line more and we don't forget what `nothing` refers to. Also easier for later refactoring.

File: src/PlaneWaveBasis.jl [hunk 8]
```diff
@@ -245,13 +246,58 @@ build_kpoints(basis::PlaneWaveBasis, kcoords) =
...
+    if symmetries === nothing
+        # TODO instead compute the group generated by ksymops (also in
+        # the other constructors). Not critical because this should
+        # not be used in this context anyway...
+        symmetries = vcat(ksymops...)
+    end
+    PlaneWaveBasis(model, Ecut,
+                   kcoords, ksymops,
+                   fft_size, variational, symmetries,
+                   kgrid, kshift, comm_kpts)
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:289]: Strange formatting ...

File: src/fft.jl [hunk 9]
```diff
@@ -101,44 +101,27 @@ end
...
-            "For a variational method, Ecut should be less than the maximal kinetic " *
-            "energy the grid supports ($max_E)"
-        )
-    else
-        # ensure no other options are set
-        @assert supersampling == 2
-        @assert !optimize_fft_size
+function compute_fft_size(model::Model{T}, Ecut, supersampling,
+                          variational, optimize_fft_size, kcoords) where {T}
+    @assert variational
```
Comment (mfherbst) [inline src/fft.jl:106]: Remove variational argument from this function and add the assertion to PWbasis constructor?

File: test/silicon_redHF.jl [hunk 10]
```diff
@@ -24,7 +24,8 @@ function run_silicon_redHF(T; Ecut=5, grid_size=15, spin_polarization=:none, kwa
     ]
     ref_etot = -5.440593269861395
 
-    fft_size = grid_size * ones(3)
+    fft_size = fill(grid_size, 3)
+    fft_size = DFTK.next_working_fft_size.(T, fft_size) # ad-hoc fix for buggy generic FFTs
```
Comment (mfherbst) [inline test/silicon_redHF.jl:28]: Actually I think asking the user to do this explicitly is better indeed. However in that case I would probably change the user-facing function to be just a call to a tuple and not require the extra broadcast.

File: src/PlaneWaveBasis.jl (hunk 2 above)
Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:160]: Yeah I thought about that but probably if you want to start playing around with stuff like that you should be able to look at the code surrounding an error message; if not, you'd better turn back anyway

**Comment by mfherbst** [comment]
I think I would already add a constructor `PlaneWaveBasis(model; Ecut=stuff, kwargs...)` if you agree to allow for a smooth transition in the future. When I record the JuliaCon workshop it would be nice to already support that pattern and don't show code examples that will break in a few weeks.

File: src/PlaneWaveBasis.jl (hunk 3 above)
Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:162]: my avowed goal is that at some point you'll give up :-p I still don't understand why you care about that, I've never seen that rule before and it makes no sense to me

File: src/PlaneWaveBasis.jl (hunk 2 above)
Comment (mfherbst) [reply src/PlaneWaveBasis.jl:160]: You could at least point out the name of the cryptic function to use for fixing this.

Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:160]: it's right there up one line from this error message. Looking up the stacktrace isn't particularly hard to do, and I don't think it's reasonable to babysit users who can't do that for this kind of advanced functionality

File: src/PlaneWaveBasis.jl (hunk 6 above)
Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:257]: because we call MPI and it might be the first time we do

File: src/PlaneWaveBasis.jl (hunk 7 above)
Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:299]: good point!

File: src/fft.jl (hunk 9 above)
Comment (antoine-levitt) [reply src/fft.jl:106]: yup good

File: src/PlaneWaveBasis.jl (hunk 2 above)
Comment (mfherbst) [reply src/PlaneWaveBasis.jl:160]: You really don't want to be nice, do you? I would consider my self an advanced user and I hate looking up what the function name was when I simply forgot to call the function. If you print it, however, I don't have to actively dig through the code next time I use this feature. I would appreciate that. I assume others, too. ;)

File: src/PlaneWaveBasis.jl (hunk 3 above)
Comment (mfherbst) [reply src/PlaneWaveBasis.jl:162]: it makes comments easier to read. python uses the same convention (and so do the established julia styles if i remember correctly)

File: src/PlaneWaveBasis.jl (hunk 6 above)
Comment (mfherbst) [reply src/PlaneWaveBasis.jl:257]: but the mpi call happens in the internal constructor, right? so there is another init call

File: src/PlaneWaveBasis.jl (hunk 7 above)
Comment (mfherbst) [reply src/PlaneWaveBasis.jl:299]: well that's one of the reasons why I dislike "true", "false" or "nothing" as positional arguments ...

File: src/PlaneWaveBasis.jl (hunk 6 above)
Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:257]: ah sorry I was not at the right line. Yup you're right

File: src/PlaneWaveBasis.jl (hunk 2 above)
Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:160]: alright, alright. It's just so ugly that we even have this that I didn't want to enshrine it in an error message, but fair enough

Comment (mfherbst) [reply src/PlaneWaveBasis.jl:160]: Yeah, but I doubt there will be a generic pure-Julia FFT any time soon. Attention pops up periodically, but then it decays when people find out how much effort went into FFTW.

Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:160]: https://github.com/JuliaApproximation/FastTransforms.jl/blob/master/src/fftBigFloat.jl looks OK; doing a simple radix-2 FFT is not hard. It's probably not super fast, but we don't really care about that

Comment (mfherbst) [reply src/PlaneWaveBasis.jl:160]: I know about that code. We tried initially and then moved away from it. Don't remember why, though. Could be that the one we use now is faster, actually.

**Comment by antoine-levitt** [comment]
OK let's do that. Thank god for the implicit X=X in kwargs in 1.6

**Comment by mfherbst** [comment]
Oh yes! I love it.
